### PR TITLE
Introduce Batch Normalization and Bypass Layers, split dnn_layer.h into layers folder

### PR DESCRIPTION
**Pull request by antonok-edm**

### CODE REVIEW

File: benchmarks/test_bwd_bn/CMakeLists.txt [hunk 1]
```diff
@@ -0,0 +1,29 @@
...
+# Add source file
+file(GLOB_RECURSE BENCHMARK_SRC_FILE RELATIVE 
+     ${CMAKE_CURRENT_SOURCE_DIR} 
+     *.cc)
+
+# Print out the source file information
+message(STATUS ${BENCHMARK_NAME} ": Benchmark source files: " ${BENCHMARK_SRC_FILE})
+
+# Add benchmark executable
+add_executable(${BENCHMARK_NAME} ${BENCHMARK_SRC_FILE})
```
Comment (shidong-ai) [inline benchmarks/test_bwd_bn/CMakeLists.txt]: Right now the source of each benchmark has been modified just a little bit. Could you modify that as well? Thanks a lot!

File: benchmarks/test_bwd_bn/test_bwd_bn.cc [hunk 2]
```diff
@@ -0,0 +1,25 @@
+#include <iostream>
+#include <gflags/gflags.h>
```
Comment (shidong-ai) [inline benchmarks/test_bwd_bn/test_bwd_bn.cc]: This is not needed anymore, we are now using #include "usage.h" instead

File: benchmarks/test_bwd_bn/test_bwd_bn.cc [hunk 3]
```diff
@@ -0,0 +1,25 @@
...
+#include "common.h"
+#include "dnnmark.h"
+
+using namespace dnnmark;
+
+DEFINE_string(config, "",
+    "The self defined DNN config file.");
+DEFINE_int32(debuginfo, 0,
+    "The debug info switch to turn on/off debug information.");
```
Comment (shidong-ai) [inline benchmarks/test_bwd_bn/test_bwd_bn.cc]: The explicit definition is not needed anymore

File: benchmarks/test_bwd_bn/test_bwd_bn.cc [hunk 4]
```diff
@@ -0,0 +1,25 @@
...
+
+int main(int argc, char **argv) {
+  google::ParseCommandLineFlags(&argc, &argv, true);
+  google::InitGoogleLogging(argv[0]);
+  FLAGS_alsologtostderr = FLAGS_debuginfo;
```
Comment (shidong-ai) [inline benchmarks/test_bwd_bn/test_bwd_bn.cc]: All the glog and gflag statement are replaced by two marco definition

File: core/include/layers/conv_layer.h [hunk 5]
```diff
@@ -0,0 +1,307 @@
...
+// 
+// THE SOFTWARE IS PROVIDED "AS IS", WITHOUT WARRANTY OF ANY KIND, EXPRESS OR
+// IMPLIED, INCLUDING BUT NOT LIMITED TO THE WARRANTIES OF MERCHANTABILITY,
+// FITNESS FOR A PARTICULAR PURPOSE AND NONINFRINGEMENT. IN NO EVENT SHALL THE
+// AUTHORS OR COPYRIGHT HOLDERS BE LIABLE FOR ANY CLAIM, DAMAGES OR OTHER
+// LIABILITY, WHETHER IN AN ACTION OF CONTRACT, TORT OR OTHERWISE, ARISING FROM,
+// OUT OF OR IN CONNECTION WITH THE SOFTWARE OR THE USE OR OTHER DEALINGS IN THE
+// SOFTWARE.
+
+#ifndef LAYERS_INCLUDE_CONV_LAYER_H_
```
Comment (shidong-ai) [inline core/include/layers/conv_layer.h]: The name of this Macro should be CORE_INCLUDE_LAYERS_CONV_LAYER_H_. It stands for the path where this file is stored

File: core/include/layers/conv_layer.h [hunk 6]
```diff
@@ -0,0 +1,307 @@
...
+// IMPLIED, INCLUDING BUT NOT LIMITED TO THE WARRANTIES OF MERCHANTABILITY,
+// FITNESS FOR A PARTICULAR PURPOSE AND NONINFRINGEMENT. IN NO EVENT SHALL THE
+// AUTHORS OR COPYRIGHT HOLDERS BE LIABLE FOR ANY CLAIM, DAMAGES OR OTHER
+// LIABILITY, WHETHER IN AN ACTION OF CONTRACT, TORT OR OTHERWISE, ARISING FROM,
+// OUT OF OR IN CONNECTION WITH THE SOFTWARE OR THE USE OR OTHER DEALINGS IN THE
+// SOFTWARE.
+
+#ifndef LAYERS_INCLUDE_CONV_LAYER_H_ 
+#define LAYERS_INCLUDE_CONV_LAYER_H_
+
```
Comment (shidong-ai) [inline core/include/layers/conv_layer.h]: Will it work without a include statement? Is it possible that you didn't remove the corresponding class in dnn_layer.h?

**Comment by shidong-ai** [comment]
Hi Anton,

I left several review comments.
For the new benchmark you added, something in CMakeLists.txt and .cc file should be modified. You can refer to the other benchmark source code for a clue.
As for the new split source file for each layer, you should modify the Macro definition according to the google code style, besides that, you didn't add an include statement to indicate where the parent class is declared and defined, which I doubt this could work. Perhaps it is because you didn't remove the old code.
Thank you very much for your work. Really appreciate!

**Comment by antonok-edm** [comment]
Sorry about all that, I didn't realize you had changed the benchmarks. I also wasn't really satisfied with the way I split the layers but I thought it had to be that way because of templates needing to be in the same file. Looks like I didn't try hard enough!
I addressed everything you commented on, let me know if anything else needs to be fixed.

File: benchmarks/test_bwd_bypass/CMakeLists.txt [hunk 7]
```diff
@@ -0,0 +1,29 @@
...
+# Add source file
+file(GLOB_RECURSE BENCHMARK_SRC_FILE RELATIVE 
+     ${CMAKE_CURRENT_SOURCE_DIR} 
+     *.cc)
+
+# Print out the source file information
+message(STATUS ${BENCHMARK_NAME} ": Benchmark source files: " ${BENCHMARK_SRC_FILE})
+
+# Add benchmark executable
+add_executable(${BENCHMARK_NAME} ${BENCHMARK_SRC_FILE})
```
Comment (shidong-ai) [inline benchmarks/test_bwd_bypass/CMakeLists.txt]: You missed this line

File: benchmarks/test_bwd_bypass/test_bwd_bn.cc [hunk 8]
```diff
@@ -0,0 +1,25 @@
+#include <iostream>
+#include <gflags/gflags.h>
```
Comment (shidong-ai) [inline benchmarks/test_bwd_bypass/test_bwd_bn.cc]: You missed this file as well. Note that the file name is incorrect.

File: benchmarks/test_fwd_bypass/CMakeLists.txt [hunk 9]
```diff
@@ -0,0 +1,29 @@
...
+# Add source file
+file(GLOB_RECURSE BENCHMARK_SRC_FILE RELATIVE 
+     ${CMAKE_CURRENT_SOURCE_DIR} 
+     *.cc)
+
+# Print out the source file information
+message(STATUS ${BENCHMARK_NAME} ": Benchmark source files: " ${BENCHMARK_SRC_FILE})
+
+# Add benchmark executable
+add_executable(${BENCHMARK_NAME} ${BENCHMARK_SRC_FILE})
```
Comment (shidong-ai) [inline benchmarks/test_fwd_bypass/CMakeLists.txt]: As above

File: benchmarks/test_fwd_bypass/test_fwd_bn.cc [hunk 10]
```diff
@@ -0,0 +1,25 @@
+#include <iostream>
+#include <gflags/gflags.h>
```
Comment (shidong-ai) [inline benchmarks/test_fwd_bypass/test_fwd_bn.cc]: As above

**Comment by shidong-ai** [comment]
No worries! You have done a great job on that! Really appreciate! :)
Sometimes it is really difficult for a person to find the mistake in the code. And this is also why code review is so very much important.
I found something else needs further modification. In your bypass layer benchmark, 1) you didn't modify the CMakeLists.txt and source code. 2) you need to change the file name as well
Thank you very much for your work!

**Comment by antonok-edm** [comment]
Should be all set!

**Pull request merged**